File: UniVote-IOT-main/Web/compare_faces.py
import sys
import json
import face_recognition

def compare_faces(stored_face_path, live_face_path):
    try:
        stored_image = face_recognition.load_image_file(stored_face_path)
        live_image = face_recognition.load_image_file(live_face_path)

        stored_encodings = face_recognition.face_encodings(stored_image)
        live_encodings = face_recognition.face_encodings(live_image)

        if len(stored_encodings) == 0 or len(live_encodings) == 0:
            return {'success': False, 'message': 'No face detected in one or both images.'}

        stored_encoding = stored_encodings[0]
        live_encoding = live_encodings[0]

        distance = face_recognition.face_distance([stored_encoding], live_encoding)[0]
        match_percentage = (1 - distance) * 100

        results = face_recognition.compare_faces([stored_encoding], live_encoding)

        if results[0]:
            return {'success': True, 'message': 'Faces match.', 'match_percentage': match_percentage}
        else:
            return {'success': False, 'message': 'Faces do not match.', 'match_percentage': match_percentage}
    except Exception as e:
        return {'success': False, 'message': str(e)}

if __name__ == "__main__":
    stored_face_path = sys.argv[1]
    live_face_path = sys.argv[2]
    result = compare_faces(stored_face_path, live_face_path)
    print(json.dumps(result))

File: UniVote-IOT-main/Web/compare_faces.php
<?php
if ($_SERVER["REQUEST_METHOD"] == "POST") {
    if (isset($_POST['storedFace']) && isset($_POST['liveFace'])) {
        $storedFace = base64_decode($_POST['storedFace']);
        $liveFace = base64_decode($_POST['liveFace']);

        // Save the images to temporary files
        $storedFacePath = tempnam(sys_get_temp_dir(), 'storedFace') . '.jpg';
        $liveFacePath = tempnam(sys_get_temp_dir(), 'liveFace') . '.jpg';

        file_put_contents($storedFacePath, $storedFace);
        file_put_contents($liveFacePath, $liveFace);

        // Call the Python script
        $output = shell_exec("python compare_faces.py $storedFacePath $liveFacePath 2>&1");

        // Debugging output
        error_log("Python script output: $output");

        // Delete temporary files
        unlink($storedFacePath);
        unlink($liveFacePath);

        if ($output === NULL) {
            echo json_encode(['success' => false, 'message' => 'Face comparison failed. No output from Python script.']);
        } else {
            $result = json_decode($output, true);
            if (json_last_error() !== JSON_ERROR_NONE) {
                error_log("JSON decode error: " . json_last_error_msg());
                echo json_encode(['success' => false, 'message' => 'Failed to decode JSON response from Python script.']);
            } else {
                if ($result['success']) {
                    if ($result['match_percentage'] < 70) {
                        $result['message'] = 'Faces do not match. Match percentage is below 70%.';
                    }
                }
                echo json_encode($result);
            }
        }
        exit;
    } else {
        echo json_encode(['success' => false, 'message' => 'Invalid input.']);
        exit;
    }
}
?>
